### PR TITLE
boards: add a new linker script that includes the .apps section

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

The `.apps` section is used as a placeholder for replacing with a compiled app binary to create a binary that includes both the kernel in the app. This is used by some boards and avoids needing to flash several times.

As a side effect, if the .elf with .apps is used to flash the kernel, this also (effectively) removes all installed apps.

This PR makes this optional rather than the default. Boards that want this can use the `kernel_layout_apps.ld` linker script.

For the stm32 discovery boards I removed the docs from the readme because those have tockloader support.

I also had to expand the build.rs code to handle relative imports outside or the board's directory. I don't think it's the prettiest code.

### Testing Strategy

Installing the microbit kernel and verifying it does not erase my apps any more.

### TODO or Help Wanted

Nothing

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by brghena — APPROVED** [review]
This makes sense. It's exactly the kind of little quality-of-life thing that is easy to not notice, but reduces friction for using Tock.

The changes look good to me.

File: boards/kernel_layout_apps.ld [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+         * the linker will ignore it when defining the segments.
+         * If less then 4 bytes, some linkers set this section to size 0
+         * and openocd fails to write it.
+         *
+         * An issue has been submitted https://github.com/raspberrypi/openocd/issues/25
+         */
+        BYTE(0xFF)
+        BYTE(0xFF)
+        BYTE(0xFF)
+        BYTE(0xFF)
```
Comment (brghena) [inline boards/kernel_layout_apps.ld:40]: @alexandruradovici Do you know whether this is still necessary?

It looks like the issue was closed a while ago: https://github.com/raspberrypi/openocd/issues/25

I guess we could just leave these four bytes as-is. Previously there was a single `BYTE(0)` here, so this isn't worse.

Comment (brghena) [reply boards/kernel_layout_apps.ld:40]: Maybe just removing the link to the issue is the right thing to do at this point.

**Review by alistair23 — APPROVED** [review]

**Comment by ppannuto** [comment]
Instead of maintaining two linker files, I believe we can just modify the elf instead? i.e.,

    arm-none-eabi-objcopy --set-section-flags .apps=NOLOAD test.elf

should prevent the section from being overwritten.

Or, perhaps more in line with the common case for tock, we mark the `.apps` section as `NOLOAD` in the linker script, and for the boards that integrate apps with the kernel image, add `arm-none-eabi-objcopy --set-section-flags .apps=LOAD test.elf`

**Comment by hudson-ayers** [comment]
> Instead of maintaining two linker files, I believe we can just modify the elf instead? i.e.,
> 
> 
> 
>     arm-none-eabi-objcopy --set-section-flags .apps=NOLOAD test.elf
> 
> 
> 
> should prevent the section from being overwritten.
> 
> 
> 
> Or, perhaps more in line with the common case for tock, we mark the `.apps` section as `NOLOAD` in the linker script, and for the boards that integrate apps with the kernel image, add `arm-none-eabi-objcopy --set-section-flags .apps=LOAD test.elf`

Is there a good way to do this without making builds only work if you build using `make`? My understanding was that We want kernels compiled using `cargo build` to work the same way as kernels compiled using that boards make target

**Comment by ppannuto** [comment]
Hah, no and yes? I was reviewing this, and I actually started by re-investigating the "how do to a conditional in a linker script" rabbit hole [where the consensus solution is apparently to run the C preprocessor over your linker script via `make` 😳 ]. Then I thought, wait, there _has_ to be a way to just edit the flags of an elf section, a quick search or three later and here we are :)

---

> Is there a good way to do this without making builds only work if you build using make?

Without reviewing the machinery, don't we already need other "extra commands" to set up the binary image for apps that will be linked for the unified flash case? That's already a bit of a hack that I think we'd rather purge from the tree wholly eventually generally...

**Comment by bradjc** [comment]
Alternative in #4169.

**Pull request closed without merging**